yoti-sdk-impl/src/main/java/com/yoti/api/client/spi/remote/HumanProfileAdapter.java

```diff
@@ -50,70 +60,110 @@ public Attribute getAttribute(String name) {
 
     @Override
     public Attribute<String> getFamilyName() {
-        return wrapped.getAttribute(AttributeConstants.HumanProfileAttributes.FAMILY_NAME, String.class);
+        return wrapped.getAttribute(HumanProfileAttributes.FAMILY_NAME, String.class);
     }
 
     @Override
     public Attribute<String> getGivenNames() {
-        return wrapped.getAttribute(AttributeConstants.HumanProfileAttributes.GIVEN_NAMES, String.class);
+        return wrapped.getAttribute(HumanProfileAttributes.GIVEN_NAMES, String.class);
     }
 
     @Override
     public Attribute<String> getFullName() {
-        return wrapped.getAttribute(AttributeConstants.HumanProfileAttributes.FULL_NAME, String.class);
+        return wrapped.getAttribute(HumanProfileAttributes.FULL_NAME, String.class);
     }
 
     @Override
     public Attribute<Date> getDateOfBirth() {
-        return wrapped.getAttribute(AttributeConstants.HumanProfileAttributes.DATE_OF_BIRTH, Date.class);
+        return wrapped.getAttribute(HumanProfileAttributes.DATE_OF_BIRTH, Date.class);
+    }
+
+    @Override
+    public List<AgeVerification> getAgeVerifications() {
+        findAllAgeVerifications();
+        return ageVerifications;
+    }
+
+    @Override
+    public AgeVerification findAgeOverVerification(int age) {
+        return findAgeVerification(HumanProfileAttributes.AGE_OVER.replace(":", ""), age);
+    }
+
+    @Override
+    public AgeVerification findAgeUnderVerification(int age) {
+        return findAgeVerification(HumanProfileAttributes.AGE_UNDER.replace(":", ""), age);
+    }
+
+    private AgeVerification findAgeVerification(String checkPerformed, int age) {
+        findAllAgeVerifications();
+        for (AgeVerification ageVerification : ageVerifications) {
+            if (ageVerification.getAgeVerified() == age && ageVerification.getCheckPerformed().equals(checkPerformed)) {
+                return ageVerification;
+            }
+        }
+        return null;
+    }
+
+    private void findAllAgeVerifications() {
+        if (ageVerifications == null) {
+            List<AgeVerification> verifications = new ArrayList<>();
+            for (Attribute<String> ageOver : wrapped.findAttributesStartingWith(HumanProfileAttributes.AGE_OVER, String.class)) {
```

Comment:
I think it's better to have a rule for each age derivation and have a processor that will apply those rules accordingly.
So we can easily scale the age verification process should there be a new derivation.
An example can be seen here:
[Age Processor](https://github.com/getyoti/yoti-php-sdk/blob/master/src/Yoti/Util/Age/Processor.php)
[Age Under|Over Processor](https://github.com/getyoti/yoti-php-sdk/blob/master/src/Yoti/Util/Age/AgeUnderOverProcessor.php)
If we have a new derivation like age range, we could have a new Processor for it, which can be added to the process list [here](https://github.com/getyoti/yoti-php-sdk/blob/master/src/Yoti/Util/Age/Processor.php#L62)

Reply:
As far as we know, we'll only ever deal with _age_over_ + _age_under_.

The plan is to wrap the _Attribute_ object in an _AgeVerification_ object.  The _AgeVerification_ object provides additional, relevant functionality - so it is the helper.  We shouldn't need to expose another helper beyond that one.  So the processing to be done is 
1. deciding what to wrap and what to wrap it in, and 
2. producing the wrapper

Pulling logic out into dedicated components is normally nice.  It's normally more testable, you can keep things focussed.  But in these SDKs we have no DI framework, so if we split things up we have to wire all the components together ourselves.  If we do it at every opportunity we'll likely end up in a mess, and in this case for _findAllAgeVerifications()_ it seems quite reasonable to keep it inside _HumanProfileAdapter_.

The thing I don't like with the _AgeProcessor_ and _Age Under|Over Processor_ is that each Processor is responsible for identifying what it matches.  So its possible to have more than one _Processor_ claiming to be the wrapper for a given attribute, and I don't think we want to go down that road.

We could maintain a mapping of attribute derivations to the classes that should wrap them, then iterate all the attributes and check the map to see if it matches.  The map keys have to be unique, so we can be sure there will only ever be one match.  That's a perfectly valid alternative to this approach.  If we do ever get more than two _age_derivations_ we could switch to that if we think its cleaner.

Reply:
Ok, either way is fine by me as long as we can easily expand this functionality.
Also I do like having ageVerifications array as a map, where the key is the attribute name and the value is the ageVerification object.

Reply:
Yep, I'm just changing that